fvm/errors/errors.go

```diff
@@ -97,3 +99,162 @@ func HandleRuntimeError(err error) error {
 	// All other errors are non-fatal Cadence errors.
 	return NewCadenceRuntimeError(runErr)
 }
+
+type prependMsgError struct {
+	msg string
+	err error
+}
+
+func (err prependMsgError) Error() string {
+	return fmt.Sprintf("%s: %v", err.msg, err.err)
+}
+
+func (err prependMsgError) Unwrap() error {
+	return err.err
+}
+
+// TODO(patrick): combine ErrorCodeChecker, Error and Failure interface.
+// We can also combine CodedError and CodedFailure once the interface is
+// unified.
+type ErrorCodeChecker interface {
+	// TODO(patrick): add Code() ErrorCode
+	// TODO(patrick): add IsFailure() bool
+
+	HasErrorCode(code ErrorCode) bool
+
+	error
+}
+
+func HasErrorCode(err error, code ErrorCode) bool {
+	var checker ErrorCodeChecker
+	return As(err, &checker) && checker.HasErrorCode(code)
+}
+
+type codedError struct {
+	code ErrorCode
+
+	// The first error in the list is the root cause.
+	errors []error
+}
+
+func wrapError(
+	code ErrorCode,
+	rootCause error,
+) codedError {
+	// TODO(patrick): check if rootCause is a failure, and if so, invert the
+	// error nesting (convert the new code into an additional error).
+	return codedError{
+		code:   code,
+		errors: []error{rootCause},
+	}
+}
+
+func newError(
+	code ErrorCode,
+	format string,
+	formatArguments ...interface{},
+) codedError {
+	return wrapError(code, fmt.Errorf(format, formatArguments...))
+}
+
+func (err codedError) Error() string {
+	if len(err.errors) == 1 {
+		return fmt.Sprintf("%v %v", err.code, err.errors[0])
+	}
+	return fmt.Sprintf(
+		"%v: %s",
+		err.code,
+		multierror.ListFormatFunc(err.errors))
+}
+
+func (err codedError) Unwrap() error {
+	return err.errors[0]
+}
+
+// This returns true if the codedError's code is set to the specified code, or
+// if any one of the codeError's errors has the error code.
+func (err codedError) HasErrorCode(code ErrorCode) bool {
+	if err.code == code {
+		return true
+	}
+
+	for _, e := range err.errors {
+		if HasErrorCode(e, code) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (err *codedError) prependMsgToRootCause(
+	format string,
+	arguments ...interface{},
+) {
+	err.errors[0] = prependMsgError{
+		msg: fmt.Sprintf(format, arguments...),
+		err: err.errors[0],
+	}
+}
+
+func (err *codedError) caused(additionalError error) {
+	err.errors = append(err.errors, additionalError)
+}
+
+type CodedError struct {
+	codedError
+}
+
+func NewCodedError(
+	code ErrorCode,
+	format string,
+	arguments ...interface{},
+) *CodedError {
+	return &CodedError{
+		codedError: newError(code, format, arguments...),
+	}
+}
+
+func (err CodedError) Code() ErrorCode {
+	return err.code
+}
+
+func (err *CodedError) PrependMsgToRootCause(
```

Comment:
Maybe just call it `Errorf`, as the behavior is very similar to `fmt.Errorf`.
```suggestion
func (err *CodedError) Errorf(
```

Reply:
Errorf returns a new error whereas this modify and return the original error.  How about Prependf instead of Errorf?

Reply:
hmm.  I think i'll drop this all together.  most of time, we immediate wrap the newly created error with fmt.Errorf.  I think pass the msg prefix directly into New<blah>Error